{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import io"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def writeEntry(text, data, f, quotes):\n",
    "    '''\n",
    "    If the data is not missing write the property to the provided file.\n",
    "\n",
    "    @text   string of the property\n",
    "    @data   data representing the property\n",
    "    @f      file to which to write the property in .mcf format\n",
    "    @quotes boolean specifying if quotes need to be added around the data entry \n",
    "          for a string\n",
    "    '''\n",
    "    if len(data) > 0  and data!=',':\n",
    "        if quotes:\n",
    "            f.write(text + '\"' + data + '\"\\n')\n",
    "    else:\n",
    "        f.write(text + data + '\\n')\n",
    "\n",
    "def addToDict(dictionary, key, value):\n",
    "    '''\n",
    "    Add value to a dictionary if the value is not empty.\n",
    "    @dictionary dict to which to add the key, value pair\n",
    "    @key        key to which to associate the value\n",
    "    @value      value to be associated with the key\n",
    "    @return     updated dictionary\n",
    "    '''\n",
    "    if len(value) > 0:\n",
    "        dictionary[key] = value\n",
    "    return(dictionary)\n",
    "\n",
    "def writeEntry2(dictionary, key, f, quotes):\n",
    "    '''\n",
    "    If the data is not missing write the property to the provided file.\n",
    "\n",
    "    @dictionary   stores key (property), value as a pair\n",
    "    @key          string of the property and the key in the dictionary\n",
    "    @f            file to which to write the property in .mcf format\n",
    "    @quotes       boolean specifying if quotes need to be added around the data\n",
    "                entry for a string\n",
    "    '''\n",
    "    if key in dictionary and quotes:\n",
    "        f.write(key + ': \"' + dictionary[key] + '\"\\n')\n",
    "    elif key in dictionary:\n",
    "        f.write(key + ': ' + dictionary[key] + '\\n')\n",
    "\n",
    "def writeEntry3(prop, value, f, quotes):\n",
    "    '''\n",
    "    If the data is not missing write the property to the provided file.\n",
    "\n",
    "    @prop     string of the property to write to the file\n",
    "    @value    string of the value to write to the file\n",
    "    @f        file to which to write the property in .mcf format\n",
    "    @quotes   boolean specifying if quotes need to be added around the data\n",
    "              entry for a string\n",
    "  '''\n",
    "    if len(value) > 0 and quotes and value!=',':\n",
    "        f.write(prop + ': \"'+ value + '\"\\n')\n",
    "    elif len(value) > 0  and value!=',':\n",
    "        f.write(prop + value + '\\n')\n",
    "        \n",
    "def writeList(dictionary, value, f):\n",
    "    '''\n",
    "    If the data is not missing write the property to the provided file.\n",
    "    Properties will be a list of values that are comma seperated. It\n",
    "    assumes the list is of dcids to be written to a property.\n",
    "\n",
    "    @dictionary   stores key (property), value as a pair\n",
    "    @value         string of the property and the key in the dictionary\n",
    "    @f            file to which to write the property in .mcf format\n",
    "    '''\n",
    "    if value in dictionary:\n",
    "        w.write(value + ': dcid:' + dictionary[value][0])\n",
    "        if len(dictionary[value]) > 0:\n",
    "            for ID in dictionary[value][1:]:\n",
    "                w.write(',dcid:' + ID)\n",
    "        w.write('\\n')\n",
    "        \n",
    "def writeList2(dictionary, value, f):\n",
    "    '''\n",
    "    If the data is not missing write the property to the provided file.\n",
    "    Properties will be a list of values that are comma seperated. It does\n",
    "    not assume that the list are dcid's to be written to a property as a\n",
    "    list.\n",
    "\n",
    "    @dictionary   stores key (property), value as a pair\n",
    "    @value         string of the property and the key in the dictionary\n",
    "    @f            file to which to write the property in .mcf format\n",
    "    '''\n",
    "    if value in dictionary:\n",
    "        w.write(value + ': ' + dictionary[value][0])\n",
    "        if len(dictionary[value]) > 0:\n",
    "            for ID in dictionary[value][1:]:\n",
    "                w.write(',' + ID)\n",
    "        w.write('\\n')\n",
    "        \n",
    "def convertToStrandEnum(strand):\n",
    "    if strand == '+':\n",
    "        return('dcid:StrandOrientationPositive')\n",
    "    if strand == '-':\n",
    "        return('dcid:StrandOrientationNegative')\n",
    "    \n",
    "def writeExonCoordinates(starts, stops, f):\n",
    "    index = 0\n",
    "    if len(starts) > 0:\n",
    "        f.write('exonCoordinates: [' + starts[0] + ' ' + stops[0] + ' Position]')\n",
    "        index += 1\n",
    "    while index < len(starts):\n",
    "        f.write(',[' + starts[index] + ' ' + stops[index] + ' Position]')\n",
    "        index += 1\n",
    "    f.write('\\n')\n",
    "\n",
    "def writeNomenclatureStatus(ns, f):\n",
    "    # convert the nomenclature status to the dcid of the corresponding enum\n",
    "    nomenclatureStatus = ''\n",
    "    if ns == 'O':\n",
    "        nomenclatureStatus = 'dcid:NomenclatureStatusOfficial'\n",
    "    elif ns == 'I':\n",
    "        nomenclatureStatus = 'dcid:NomenclatureStatusInterim'\n",
    "    elif line[12] == '-':\n",
    "        ns = 'dcid:NomenclatureStatusNCBIsupplied'\n",
    "    writeEntry3('nomenclatureStatus: ', nomenclatureStatus, f, False)\n",
    "    \n",
    "def writeTypeOfGene(gt, f):\n",
    "    geneType = ''\n",
    "    if gt == 'biological-region':\n",
    "        geneType = 'dcid:TypeOfGeneBiologicalRegion'\n",
    "    elif gt == 'ncRNA':\n",
    "        geneType = 'dcid:TypeOfGenencRNA'\n",
    "    elif gt == 'other':\n",
    "        geneType = 'dcid:TypeOfGeneOther'\n",
    "    elif gt == 'protein-coding':\n",
    "        geneType = 'dcid:TypeOfGeneProteinCoding'\n",
    "    elif gt == 'pseudo':\n",
    "        geneType = 'dcid:TypeOfGenePseudo'\n",
    "    elif gt == 'rRNA':\n",
    "        geneType = 'dcid:TypeOfGenerRNA'\n",
    "    elif gt == 'scRNA':\n",
    "        geneType = 'dcid:TypeOfGenescRNA'\n",
    "    elif gt == 'snRNA':\n",
    "        geneType = 'dcid:TypeOfGenesnRNA'\n",
    "    elif gt == 'snoRNA':\n",
    "        geneType = 'dcid:TypeOfGenesnoRNA'\n",
    "    elif gt == 'tRNA':\n",
    "        geneType = 'dcid:TypeOfGenetRNA'\n",
    "    elif gt == 'miscRNA':\n",
    "        geneType = 'dcid:TypeOfGenemiscRNA'\n",
    "    elif gt == 'unknown':\n",
    "        geneType = 'dcid:TypeOfGeneUnknown'\n",
    "    else:\n",
    "        print(gt)\n",
    "    writeEntry3('typeOfGene: ', geneType, f, False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of UCSCTranscript instances: 226811\n"
     ]
    }
   ],
   "source": [
    "# set assembly name\n",
    "assembly = 'hg38'\n",
    "\n",
    "# set species name\n",
    "species = 'Homo_sapiens'\n",
    "species_abv = 'hs'\n",
    "\n",
    "# set file path\n",
    "path = '/Users/spiekos/Documents/files+mcf_creation_code/gene_coordinates'\n",
    "\n",
    "# save info from RefSeq file for Genes\n",
    "dict_gene = {}\n",
    "\n",
    "# write the RefSeq instances to the *_RefSeq.mcf output file\n",
    "refSeq = assembly + '_gene_coordinates.txt'\n",
    "refSeq_file = os.path.join(path, refSeq)\n",
    "r = open(refSeq_file, 'r')\n",
    "transcript_output_file = assembly + '_RNATranscript.mcf'\n",
    "refSeq_mcf_file = os.path.join(path, 'RNATranscript_Gene_mcfs', transcript_output_file)\n",
    "w = open(refSeq_mcf_file, 'w')\n",
    "next(r)\n",
    "count = 0\n",
    "for line in r:\n",
    "    line = line.strip('\\r\\n').strip('\\\\').split('\\t')\n",
    "    count += 1\n",
    "    dcid = 'bio/'  + assembly + '_' + line[0].replace(',', '')\n",
    "    w.write('Node: dcid:' + dcid + '\\n')\n",
    "    w.write('ofSpecies: dcid:bio/' + species_abv + '\\n')\n",
    "    w.write('name: \"' + line[0].replace(',', '') + '\"\\n')\n",
    "    w.write('inGenomeAssembly: dcid:bio/' + assembly + '\\n')\n",
    "    w.write('strandOrientation: '+ convertToStrandEnum(line[2].replace(',', ''))  + '\\n')\n",
    "    w.write('inChromosome: dcid:bio/' + assembly + '_' + line[1].replace(',', '')  + '\\n')\n",
    "    w.write('transcriptionCoordinates: [' + line[3].replace(',', '')  + ' ' + line[4].replace(',', '')  + ' Position]\\n')\n",
    "    w.write('codingCoordinates: [' + line[5].replace(',', '')  + ' ' + line[6].replace(',', '')  + ' Position]\\n')\n",
    "    writeEntry('exonFrame: ', line[7].replace(',', '') , w, False)\n",
    "    writeExonCoordinates(line[8].split(',')[:-1], line[9].split(',')[:-1], w)\n",
    "    if species_abv == 'hs':\n",
    "        writeEntry('makesProtein: ', 'dcid:bio/UniProt_' + line[10].replace(',', '') + '_HUMAN', w, True)\n",
    "    else:\n",
    "        writeEntry('makesProtein: ', line[10].replace(',', '') , w, True)\n",
    "    writeEntry('geneSymbol: ', line[12].replace(',', '') , w, True)\n",
    "    writeEntry('refSeqID: ', line[13].replace(',', '') , w, True)\n",
    "    w.write('typeOf: schema:RNATranscript\\n')\n",
    "    w.write('\\n')\n",
    "    gene_name = line[12].replace(',', '')\n",
    "    if gene_name not in dict_gene.keys():\n",
    "        dict_gene[gene_name] = {'strandOrientation': line[2].replace(',', '')}\n",
    "        dict_gene[gene_name]['hasRNATranscript'] = [dcid]\n",
    "        dict_gene[gene_name]['inChromosome'] = line[1].replace(',', '')\n",
    "    else:\n",
    "        dict_gene[gene_name]['hasRNATranscript'].append(dcid)\n",
    "    if len(line[13]) > 0 and line[13] != ',':\n",
    "        refSeq = line[13].strip(',')\n",
    "        if 'refSeqID' not in dict_gene[gene_name].keys():\n",
    "            dict_gene[gene_name]['refSeqID'] = [refSeq]\n",
    "        elif line[13] not in dict_gene[gene_name]['refSeqID']:\n",
    "            dict_gene[gene_name]['refSeqID'].append(refSeq)\n",
    "r.close()\n",
    "w.close()\n",
    "print('Number of UCSCTranscript instances: ' + str(count))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# keep track of recorded gene - if duplicated entries, take the largest region \n",
    "dict_recorded_genes = {}\n",
    "\n",
    "# write the UCSC gene file to the dict_gene\n",
    "gene_file = assembly + '_known_canonical_gene_coordinates.txt'\n",
    "UCSC_gene_file = os.path.join(path, gene_file)\n",
    "r = open(UCSC_gene_file, 'r')\n",
    "next(r)\n",
    "for line in r:\n",
    "    line = line.strip('\\r\\n').strip('\\\\').split('\\t')\n",
    "    gene_size = int(line[2]) - int(line[1])\n",
    "    gene_name = line[5].replace(',', '')\n",
    "    if gene_name not in dict_recorded_genes:\n",
    "        dict_recorded_genes[gene_name] = gene_size\n",
    "    if gene_size < dict_recorded_genes[gene_name]:\n",
    "        continue\n",
    "    if gene_name not in dict_gene:\n",
    "        dict_gene[gene_name] = {}\n",
    "    dict_recorded_genes[gene_name] = gene_size\n",
    "    addToDict(dict_gene[gene_name], 'inChromosome', line[0])\n",
    "    addToDict(dict_gene[gene_name], 'cannonicalStart', line[1])\n",
    "    addToDict(dict_gene[gene_name], 'cannonicalStop', line[2])\n",
    "    addToDict(dict_gene[gene_name], 'mRNA', line[4].replace(',', ''))\n",
    "    addToDict(dict_gene[gene_name], 'geneSymbol', line[5].replace(',', ''))\n",
    "    addToDict(dict_gene[gene_name], 'ncbiProteinAccessionNumber', line[7].replace(',', ''))\n",
    "    addToDict(dict_gene[gene_name], 'description', line[8])\n",
    "    if len(line[6]) > 1:\n",
    "        if 'refSeqID' not in dict_gene[gene_name].keys():\n",
    "            dict_gene[gene_name]['refSeqID'] = [line[6].replace(',', '')]\n",
    "        elif line[6] not in dict_gene[gene_name]['refSeqID']:\n",
    "            dict_gene[gene_name]['refSeqID'].append(line[6].replace(',', ''))\n",
    "r.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of Gene instances: 58683\n"
     ]
    }
   ],
   "source": [
    "# write the gene nodes to a file in mcf format\n",
    "gene_output_file = assembly + '_gene.mcf'\n",
    "gene_mcf_file = os.path.join(path, 'RNATranscript_Gene_mcfs', gene_output_file)\n",
    "w = open(gene_mcf_file, 'w')\n",
    "count = 0\n",
    "for key in dict_gene:\n",
    "    count += 1\n",
    "    dcid = 'bio/' + assembly + '_' + key\n",
    "    w.write('Node: dcid:' + dcid + '\\n')\n",
    "    w.write('name: \"' + key + '\"\\n')\n",
    "    w.write('ofSpecies: dcid:bio/' + species_abv + '\\n')\n",
    "    w.write('inGenomeAssembly: dcid:bio/' + assembly + '\\n')\n",
    "    w.write('inChromosome: dcid:bio/' + assembly + '_' + dict_gene[key]['inChromosome'] + '\\n')\n",
    "    if 'cannonicalStart' in dict_gene[key] and 'cannonicalStop' in dict_gene[key]:\n",
    "        w.write('genomicCoordinates: [' + dict_gene[key]['cannonicalStart'] + ' ' + \\\n",
    "                dict_gene[key]['cannonicalStop'] + ' Position]\\n')\n",
    "    if 'strandOrientation' in dict_gene[key]:\n",
    "        w.write('strandOrientation: '+ convertToStrandEnum(dict_gene[key]['strandOrientation']) + '\\n')\n",
    "    w.write('geneSymbol: \"' + key + '\"\\n')\n",
    "    writeEntry2(dict_gene[key], 'mRNA', w, True)\n",
    "    writeList2(dict_gene[key], 'refSeqID', w)\n",
    "    writeList(dict_gene[key], 'hasRNATranscript', w)\n",
    "    writeEntry2(dict_gene[key], 'ncbiProteinAccessionNumber', w, True)\n",
    "    writeEntry2(dict_gene[key], 'description', w, True)\n",
    "    w.write('typeOf: schema:Gene\\n')\n",
    "    w.write('\\n')\n",
    "w.close()\n",
    "print('Number of Gene instances: ' + str(count))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write the NCBI gene file to dict_gene_ncbi\n",
    "dict_gene_ncbi = {}\n",
    "NCBI_gene_file = species + '.gene_info'\n",
    "r = open(NCBI_gene_file, 'r')\n",
    "next(r)\n",
    "for line in r:\n",
    "    line = line.strip('\\r\\n').strip('\\\\').split('\\t')\n",
    "    if line[2] not in dict_gene:\n",
    "        dict_gene_ncbi[line[2]] = {'inChromosome': 'chr' + line[6]}\n",
    "        dict_gene_ncbi[line[2]]['description'] = line[8].strip('-')\n",
    "    elif line[2] not in dict_gene_ncbi:\n",
    "        dict_gene_ncbi[line[2]] = {}\n",
    "    addToDict(dict_gene_ncbi[line[2]], 'ncbiTaxonID', line[0].strip('-'))\n",
    "    addToDict(dict_gene_ncbi[line[2]], 'ncbiGeneID', line[1].strip('-'))\n",
    "    addToDict(dict_gene_ncbi[line[2]], 'ncbiLocusTag', line[3].strip('-'))\n",
    "    addToDict(dict_gene_ncbi[line[2]], 'mapLocation', line[7].strip('-'))\n",
    "    addToDict(dict_gene_ncbi[line[2]], 'typeOfGene', line[9].strip('-'))\n",
    "    addToDict(dict_gene_ncbi[line[2]], 'fullName', line[11].strip('-'))\n",
    "    addToDict(dict_gene_ncbi[line[2]], 'modificationDate', line[14].strip('-'))\n",
    "    addToDict(dict_gene_ncbi[line[2]], 'nomenclatureStatus', line[12])\n",
    "    HGNC = ''\n",
    "    mim = ''\n",
    "    ensembl = ''\n",
    "    otherNames = line[2].split('|')\n",
    "    for name in otherNames:\n",
    "      # split up the alternative names to correspond to the correct property\n",
    "        n = name.split(':')\n",
    "        if name.startswith('HGNC'):\n",
    "            HGNC = n[-1]\n",
    "        elif name.startswith('mim'):\n",
    "            MIM = n[-1]\n",
    "        elif name.startswith('Ensembl'):\n",
    "            ensembl = n[-1]\n",
    "        elif name == line[2]:\n",
    "            continue\n",
    "        else:\n",
    "            print(name)\n",
    "        addToDict(dict_gene_ncbi[line[2]], 'hgncID', HGNC)\n",
    "        addToDict(dict_gene_ncbi[line[2]], 'mimID', mim)\n",
    "        addToDict(dict_gene_ncbi[line[2]], 'ensemblID', ensembl)\n",
    "r.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of Gene instances: 61559\n"
     ]
    }
   ],
   "source": [
    "# write the gene nodes to a file in mcf format\n",
    "gene_ncbi_output_file = assembly + '_gene_ncbi.mcf'\n",
    "gene_mcf_file = os.path.join(path, 'ncbi_Gene_mcfs', gene_ncbi_output_file)\n",
    "w = open(gene_mcf_file, 'w')\n",
    "count = 0\n",
    "for key in dict_gene_ncbi:\n",
    "    count += 1\n",
    "    dcid = 'bio/' + assembly + '_' + key\n",
    "    w.write('Node: dcid:' + dcid + '\\n')\n",
    "    w.write('name: \"' + key + '\"\\n')\n",
    "    w.write('ofSpecies: dcid:bio/' + species_abv + '\\n')\n",
    "    w.write('inGenomeAssembly: dcid:bio/' + assembly + '\\n')\n",
    "    if 'inChromosome' in dict_gene_ncbi[key]:\n",
    "        w.write('inChromosome: dcid:bio/' + assembly + '_' + dict_gene_ncbi[key]['inChromosome'] + '\\n')\n",
    "    writeEntry2(dict_gene_ncbi[key], 'mapLocation', w, True)\n",
    "    writeTypeOfGene(dict_gene_ncbi[key]['typeOfGene'], w)\n",
    "    w.write('geneSymbol: \"' + key + '\"\\n')\n",
    "    writeEntry2(dict_gene_ncbi[key], 'ncbiTaxonID', w, False)\n",
    "    writeEntry2(dict_gene_ncbi[key], 'ncbiGeneID', w, False)\n",
    "    writeEntry2(dict_gene_ncbi[key], 'ncbiLocusTag', w, True)\n",
    "    writeEntry2(dict_gene_ncbi[key], 'mimID', w, True)\n",
    "    writeEntry2(dict_gene_ncbi[key], 'hgncID', w, True)\n",
    "    writeEntry2(dict_gene_ncbi[key], 'ensemblID', w, True)\n",
    "    writeEntry2(dict_gene_ncbi[key], 'fullName', w, True)\n",
    "    writeNomenclatureStatus(dict_gene_ncbi[key]['nomenclatureStatus'], w)\n",
    "    writeEntry2(dict_gene_ncbi[key], 'modificationDate', w, False)\n",
    "    writeEntry2(dict_gene_ncbi[key], 'description', w, True)\n",
    "    w.write('typeOf: schema:Gene\\n')\n",
    "    w.write('\\n')\n",
    "w.close()\n",
    "print('Number of Gene instances: ' + str(count))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
